File: packages/runtime-rs/src/cid.rs
//! Content-identifier (CID) computation — mirror of `src/cid.js`.
//!
//! Format: `cidv1:sha256:<64-hex>`. Computed deterministically from a
//! manifest's `hashes` block:
//!
//! ```text
//! canonical_json({
//!   digest: "sha256",
//!   parts: {
//!     model_pointer: <hex>,
//!     recipes_json:  <hex>,
//!     lora_bin:      <hex>,
//!     index_bin:     <hex>,
//!     evals_json:    <hex>,
//!   }
//! }) → sha256 → "cidv1:sha256:<hex>"
//! ```
//!
//! Same inputs always produce the same CID; any byte difference in any of
//! the five per-file hashes changes the CID. The CID is independent of the
//! receipt signature, K-score, and tenant secret rotation.

use crate::canonical::canonical_json;
use crate::error::Error;
use crate::manifest::ManifestHashes;
use sha2::{Digest, Sha256};

/// The CID version string this crate emits.
pub const CID_VERSION: &str = "cidv1";

/// The digest name this crate emits.
pub const CID_DIGEST: &str = "sha256";

/// Compute a CID from a `ManifestHashes` block.
pub fn cid_from_manifest_hashes(h: &ManifestHashes) -> Result<String, Error> {
    validate_hex64("model_pointer", &h.model_pointer)?;
    validate_hex64("recipes_json", &h.recipes_json)?;
    validate_hex64("lora_bin", &h.lora_bin)?;
    validate_hex64("index_bin", &h.index_bin)?;
    validate_hex64("evals_json", &h.evals_json)?;

    let v = serde_json::json!({
        "digest": CID_DIGEST,
        "parts": {
            "model_pointer": h.model_pointer,
            "recipes_json":  h.recipes_json,
            "lora_bin":      h.lora_bin,
            "index_bin":     h.index_bin,
            "evals_json":    h.evals_json,
        }
    });
    let canon = canonical_json(&v);
    let mut hasher = Sha256::new();
    hasher.update(canon.as_bytes());
    let digest = hex::encode(hasher.finalize());
    Ok(format!("{}:{}:{}", CID_VERSION, CID_DIGEST, digest))
}

/// Verify a CID string against a `ManifestHashes` block. Returns `true` iff
/// the recomputed CID matches, `false` otherwise.
pub fn verify_cid(cid: &str, hashes: &ManifestHashes) -> bool {
    match cid_from_manifest_hashes(hashes) {
        Ok(computed) => computed == cid,
        Err(_) => false,
    }
}

/// Components of a parsed CID — version, digest, hex.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParsedCid {
    /// CID version (e.g. `cidv1`).
    pub version: String,
    /// Digest algorithm (e.g. `sha256`).
    pub digest: String,
    /// Hex-encoded digest body.
    pub hex: String,
}

/// Parse a CID string. Returns `None` on malformed input.
pub fn parse_cid(s: &str) -> Option<ParsedCid> {
    let parts: Vec<&str> = s.splitn(3, ':').collect();
    if parts.len() != 3 {
        return None;
    }
    if !parts[0].starts_with("cidv") || parts[0].len() < 5 {
        return None;
    }
    if parts[1].is_empty() || !parts[1].chars().all(|c| c.is_ascii_alphanumeric() || c == '-') {
        return None;
    }
    if parts[2].is_empty() || !parts[2].chars().all(|c| c.is_ascii_hexdigit()) {
        return None;
    }
    Some(ParsedCid {
        version: parts[0].to_string(),
        digest: parts[1].to_string(),
        hex: parts[2].to_string(),
    })
}

fn validate_hex64(field: &str, s: &str) -> Result<(), Error> {
    if s.len() != 64 || !s.chars().all(|c| c.is_ascii_hexdigit() && !c.is_ascii_uppercase()) {
        return Err(Error::MalformedManifest(format!(
            "hashes.{} must be 64-char lowercase hex sha256",
            field
        )));
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn h(byte: char) -> String {
        std::iter::repeat(byte).take(64).collect()
    }

    fn fixture() -> ManifestHashes {
        ManifestHashes {
            model_pointer: h('a'),
            recipes_json: h('b'),
            lora_bin: h('c'),
            index_bin: h('d'),
            evals_json: h('e'),
        }
    }

    #[test]
    fn produces_cidv1_sha256() {
        let cid = cid_from_manifest_hashes(&fixture()).unwrap();
        assert!(cid.starts_with("cidv1:sha256:"));
        assert_eq!(cid.len(), "cidv1:sha256:".len() + 64);
    }

    #[test]
    fn deterministic() {
        let a = cid_from_manifest_hashes(&fixture()).unwrap();
        let b = cid_from_manifest_hashes(&fixture()).unwrap();
        assert_eq!(a, b);
    }

    #[test]
    fn changes_when_any_input_changes() {
        let base = cid_from_manifest_hashes(&fixture()).unwrap();
        let mut m = fixture();
        m.evals_json = h('f');
        assert_ne!(cid_from_manifest_hashes(&m).unwrap(), base);
    }

    #[test]
    fn rejects_malformed_hex() {
        let mut m = fixture();
        m.evals_json = "too-short".to_string();
        assert!(cid_from_manifest_hashes(&m).is_err());
    }

    #[test]
    fn verify_cid_passes_on_match() {
        let cid = cid_from_manifest_hashes(&fixture()).unwrap();
        assert!(verify_cid(&cid, &fixture()));
    }

    #[test]
    fn verify_cid_fails_on_mismatch() {
        assert!(!verify_cid(
            "cidv1:sha256:0000000000000000000000000000000000000000000000000000000000000000",
            &fixture()
        ));
    }

    #[test]
    fn parse_cid_valid() {
        let cid = cid_from_manifest_hashes(&fixture()).unwrap();
        let parsed = parse_cid(&cid).unwrap();
        assert_eq!(parsed.version, "cidv1");
        assert_eq!(parsed.digest, "sha256");
    }

    #[test]
    fn parse_cid_invalid() {
        assert!(parse_cid("").is_none());
        assert!(parse_cid("cidv1:sha256:").is_none());
        assert!(parse_cid("bad").is_none());
        assert!(parse_cid("cidv1:sha256:XYZ").is_none());
    }
}

File: src/cid.js
// Content-identifier (CID) for .kolm artifacts.
//
// A CID is a deterministic, content-addressed identifier that any verifier
// can recompute from the artifact bytes alone - no server, no registry, no
// authentication. Two artifacts with identical content produce the same CID;
// any byte difference (even a re-zipped file order) produces a different CID.
//
// Format: `cidv1:sha256:<64-hex>` (multiformat-inspired prefix so future
// digest algorithms can be added without breaking back-compat).
//
// CID inputs (canonical sorted-key JSON over the manifest's `hashes` block):
//   - model_pointer - sha256 of model.gguf bytes
//   - recipes_json - sha256 of recipes.json bytes
//   - lora_bin - sha256 of lora.bin bytes
//   - index_bin - sha256 of index.sqlite-vec bytes
//   - evals_json - sha256 of evals.json bytes
//
// The K-score, receipt, and signature are deliberately NOT in the CID. Those
// are the *seal* on the bundle; the CID is the *identity* of the bundle. A
// receipt re-signed with a rotated tenant secret still describes the same
// underlying CID. The artifact_hash on the manifest is a flat sha256 over
// the same inputs; the CID prefixes it for future multi-hash compatibility.
//
// Deduplication: two compile jobs that yield the same CID (same task spec,
// same recipes, same evals, same base model pointer) can be deduped by the
// registry to a single artifact row. The audit log is never deduped - every
// compile remains a distinct event.

import crypto from 'node:crypto';

const CID_VERSION = 'cidv1';
const DEFAULT_DIGEST = 'sha256';
const REQUIRED_PARTS = ['model_pointer', 'recipes_json', 'lora_bin', 'index_bin', 'evals_json'];

export function canonicalJson(v) {
  if (v === null || typeof v !== 'object') return JSON.stringify(v);
  if (Array.isArray(v)) return '[' + v.map(canonicalJson).join(',') + ']';
  const k = Object.keys(v).sort();
  return '{' + k.map(x => JSON.stringify(x) + ':' + canonicalJson(v[x])).join(',') + '}';
}

function sha256Hex(input) {
  return crypto.createHash('sha256').update(input).digest('hex');
}

// Compute a CID from a manifest's `hashes` block. Pass the manifest object's
// hashes - not the receipt - so the CID is independent of signing key
// rotation and remains stable across re-seals of identical content.
export function cidFromManifestHashes(hashes) {
  if (!hashes || typeof hashes !== 'object') {
    throw new Error('cidFromManifestHashes: hashes object required');
  }
  for (const k of REQUIRED_PARTS) {
    if (typeof hashes[k] !== 'string' || !/^[0-9a-f]{64}$/.test(hashes[k])) {
      throw new Error(`cidFromManifestHashes: hashes.${k} must be a 64-char hex sha256`);
    }
  }
  const canonical = canonicalJson({
    digest: DEFAULT_DIGEST,
    parts: {
      model_pointer: hashes.model_pointer,
      recipes_json:  hashes.recipes_json,
      lora_bin:      hashes.lora_bin,
      index_bin:     hashes.index_bin,
      evals_json:    hashes.evals_json,
    },
  });
  const digest = sha256Hex(canonical);
  return `${CID_VERSION}:${DEFAULT_DIGEST}:${digest}`;
}

// Verify a CID against a manifest's hashes block - symmetric helper for the
// verifier. Returns true when the CID matches; false otherwise (never throws).
export function verifyCidAgainstManifestHashes(cid, hashes) {
  try {
    return cidFromManifestHashes(hashes) === cid;
  } catch {
    return false;
  }
}

// Parse a CID string into its components. Returns null on malformed input
// so callers can branch without try/catch.
export function parseCid(cid) {
  if (typeof cid !== 'string') return null;
  const m = /^(cidv\d+):([a-z0-9-]+):([0-9a-f]+)$/.exec(cid);
  if (!m) return null;
  return { version: m[1], digest: m[2], hex: m[3] };
}

// Strict-validate a CID format string. True iff the encoding is well-formed
// (independent of whether any underlying content matches).
export function isValidCidFormat(cid) {
  const p = parseCid(cid);
  if (!p) return false;
  if (p.version !== CID_VERSION) return false;
  if (p.digest !== DEFAULT_DIGEST) return false;
  if (p.hex.length !== 64) return false;
  return true;
}

// Short-CID for display: cidv1:sha256:abc123…def456 (first 6 + last 6 hex).
// Useful in CLI tables and dashboards where the full 64-hex digest is noise.
export function shortCid(cid) {
  const p = parseCid(cid);
  if (!p) return cid;
  if (p.hex.length <= 14) return cid;
  return `${p.version}:${p.digest}:${p.hex.slice(0, 6)}…${p.hex.slice(-6)}`;
}

export const CID_SPEC = {
  version: CID_VERSION,
  digest: DEFAULT_DIGEST,
  parts: REQUIRED_PARTS,
};
